src/charm.py

```diff
@@ -140,14 +140,49 @@ def _check_leader(self):
             self.logger.info("Not a leader, skipping setup")
             raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
-    def _deploy_k8s_resources(self) -> None:
-        """Deploys K8S resources."""
+    def _check_and_report_k8s_conflict(self, error):
+        """Check for conflict and report it.
+
+        Returns:
+          True if error status code is 409 (conflict).
+          False otherwise.
+        """
+        if error.status.code == 409:
+            self.unit.status = MaintenanceStatus("Force applying K8S resources")
+            self.logger.warning(f"{str(error)}")
+            return True
+        return False
+
+    def _apply_k8s_resources(self, force_conflicts: bool = False) -> None:
+        """Applies K8S resources.
+
+        Args:
+            force_conflicts (bool): *(optional)* Will "force" apply requests causing conflicting
+                                    fields to change ownership to the field manager used in this
+                                    charm.
+                                    NOTE: This will only be used if initial regular apply() fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8S resources")
         try:
-            self.unit.status = MaintenanceStatus("Creating K8S resources")
             self.k8s_resource_handler.apply()
+        except ApiError as error:
```

Comment:
I understand having two try/except statements here is just for the sake of making a distinction between applying CRDs and other k8s resources, but I wonder if you explored the option of just having one. Seems to me we have repeated the code, just changed the logging msg. This should not be a blocker, IMO, just something nice to have.

Reply:
Yes, the reason is exactly what you described. If it fails on K8S apply, there will be no warning and logging for the CRD and we will go an apply with force conflict resolution. There will be no logging/warning about conflicts and forced apply for the CRD.

Reply:
I even tried to use single handler for both CRD and K8S resources, but since those are in two separate files, apply has the same behaviour: if it fails on K8S, CRDs are not even tried, thus loosing error/log message about conflicts for CRDs.

Reply:
Sure, we can revisit the code later, should not be a blocker.